src/proto/connection.rs

```diff
@@ -282,7 +282,13 @@ where
                 State::Closing(reason, initiator) => {
                     tracing::trace!("connection closing after flush");
                     // Flush/shutdown the codec
-                    ready!(self.codec.shutdown(cx))?;
+                    if let Err(e) = ready!(self.codec.shutdown(cx)) {
+                        // If the error kind is NotConnected then ignore that
+                        // since it means the connection is already shutdown.
+                        if e.kind() != io::ErrorKind::NotConnected {
+                            return Poll::Ready(Err(e.into()));
+                        }
```

Comment:
what do you think about logging something at TRACE here, so if a user is looking at very verbose logs, they would be able to determine that an error was ignored here?
```suggestion
                        } else {
                            tracing::trace!(
                                "ignoring NotConnected error \
                                (the connection has already closed)"
                           );
                       }
```